File: lc_soln/0042_trapping_rain_water/solution.cs
public class Solution {
    public int Trap(int[] height) {
        int n = height.Length;
        if (n < 3) {
            return 0;
        }

        int l = 0;
        int r = n - 1;
        int lh = height[l];
        int rh = height[r];
        int result = 0;
        while (l < r) {
            if (lh < rh) {
                ++l;
                lh = int.Max(lh, height[l]);
                result += lh - height[l];
            } else {
                --r;
                rh = int.Max(rh, height[r]);
                result += rh - height[r];
            }
        }
        return result;
    }
}

File: lc_soln/0042_trapping_rain_water/solution.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        if n < 3:
            return 0
        l = 0
        r = n - 1
        lh = height[l]
        rh = height[r]
        result = 0
        while l < r:
            if lh < rh:
                l += 1
                lh = max(lh, height[l])
                result += lh - height[l]
            else:
                r -= 1
                rh = max(rh, height[r])
                result += rh - height[r]
        return result


def main(): ...


if __name__ == "__main__":
    main()
